internal/consensus/state.go

```diff
@@ -2254,7 +2254,54 @@ func (cs *State) tryAddVote(vote *types.Vote, peerID p2p.ID) (bool, error) {
 	return added, nil
 }
 
+// addVote adds a vote to the consensus state. It performs various checks and updates the state accordingly.
 func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error) {
+	cs.logVoteDetails(vote)
+
+	if cs.isLateVote(vote) {
+		cs.metrics.MarkLateVote(vote.Type)
+	}
+
+	// A precommit for the previous height?
+	// These come in while we wait timeoutCommit
+	if cs.isPrecommitForPreviousHeight(vote) {
+		return cs.handlePrecommitForPreviousHeight(vote)
+	}
+
+	// Height mismatch is ignored.
+	// Not necessarily a bad peer, but not favorable behavior.
+	if cs.isHeightMismatch(vote) {
+		cs.Logger.Debug("vote ignored and not added", "vote_height", vote.Height, "cs_height", cs.Height, "peer", peerID)
+		return false, nil
+	}
+
+	// Check to see if the chain is configured to extend votes.
+	if err := cs.verifyVoteExtension(vote, peerID); err != nil {
+		return false, err
+	}
+
+	added, err = cs.Votes.AddVote(vote, peerID, cs.state.ConsensusParams.Feature.VoteExtensionsEnabled(vote.Height))
+	if !added {
+		cs.handleDuplicateVote(err)
+		return false, err
+	}
+
+	cs.updateMetricsAndEvents(vote)
+
+	switch vote.Type {
+	case types.PrevoteType:
+		cs.handlePrevote(vote)
+	case types.PrecommitType:
+		cs.handlePrecommit(vote)
+	default:
+		panic(fmt.Sprintf("unexpected vote type %v", vote.Type))
+	}
+
+	return true, nil
+}
+
+// logVoteDetails logs the details of a vote.
+func (cs *State) logVoteDetails(vote *types.Vote) {
```

Comment:
I don't see the point of moving this to a method.

Reply:
Moving this out makes a lot of sense to me.  It takes something that isn't "core" and moves it out.  What am I missing?

Reply:
unresolve if you really want this one changed.  Some of the others were very good catches, thank you!